### PR TITLE
Add mpp_timeout and invalid_onion_payload descriptions

### LINKED ISSUE #1129: Need to support decoding more onion types
I received an onion failure after an MPP part timed out which log reported as unknown, i presume we're just out of date compared to BOLTs:

`2021-10-17 20:33:13.537 WARN  [lightning::ln::onion_utils:484] Onion Error[from ...: (0x17)] Unknown`

### PR DESCRIPTION
**Pull request by dunxen**
Fixes #1129 and gets us up to date with onion error messages types at the time of writing.

There is possibly extra data associated with `invalid_onion_payload` we could use for debugging:

From BOLT 4, type `PERM|22`:

> If the failure can be narrowed down to a specific tlv type in the payload, the erring node may include that type and its byte offset in the decrypted byte stream

I'm not sure of specific extra handling that might needed for these two error types.

### CODE REVIEW

**Review by ariard — COMMENTED** [review]
ACK 6069c7d

> I'm not sure of specific extra handling that might needed for these two error types.

I think in `decode_update_add_htlc_onion`, at OnionHopData deser we could have a newer DecodeError::UnknownTLV with a return_err!() $data set to the unknown type field ?

**Review by TheBlueMatt — COMMENTED** [review]
Sorry for the delay here. Also need to update `process_onion_failure` in `lightning/src/ln/onion_utils.rs`, specifically I believe at least `23` implies `payment_failed` should be set (as this particular HTLC/path did *not* fail, even if others did). This may require some refactoring as we need to separate the concepts of "payment is retryable" from "this HTLC was failed by the recipient for some reason other than a node along the path was malicious" which are currently considered exactly opposite, but with `23` they aren't anymore.

**Comment by dunxen** [comment]
Will probably get to this at the end of the week. Thanks for the guidance. I'll ask some more questions if I need to :)

**Comment by dunxen** [comment]
> Will probably get to this at the end of the week. Thanks for the guidance. I'll ask some more questions if I need to :)

Sorry, been meaning to get to this. Have company summit shenanigans this week.

**Comment by dunxen** [comment]
Finally have time to get back to this today and my Mondays are open going forward!

**Comment by dunxen** [comment]
Haven't touched `23` handling yet. Just checking if I'm on the right track with `UnknownTLV`.

File: lightning/src/ln/channelmanager.rs [hunk 1]
```diff
@@ -1765,6 +1765,11 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 						msgs::DecodeError::UnknownRequiredFeature|
 						msgs::DecodeError::InvalidValue|
 						msgs::DecodeError::ShortRead => 0x4000 | 22, // invalid_onion_payload
+						msgs::DecodeError::UnknownTLV(tlv_type) => {
+							let tlv_type = BigSize(tlv_type);
+							let mut v = Vec::with_capacity(tlv_type.serialized_length());
+							return_err!("Unable to decode our hop data", 0x4000 | 22, if let Ok(_) = tlv_type.write(&mut v) { &v[..] } else { &[0;0] });
```
Comment (TheBlueMatt) [inline lightning/src/ln/channelmanager.rs:1771]: Do we need to bother here? Spec says informing the sender of which type was unknown is optional and supporting it means some amount of effort on the TLV-decode end. We can add it if you want, but supporting sending a new onion error type seems unrelated to this PR (which is about handling a new onion error type).

Comment (dunxen) [reply lightning/src/ln/channelmanager.rs:1771]: Yeah then I'll rather leave it for another PR and just focus on the handling :)

**Comment by dunxen** [comment]
Sorry for the delay, have been spending time digging into some BOLT 4 stuff for more background.

> This may require some refactoring as we need to separate the concepts of "payment is retryable" from "this HTLC was failed by the recipient for some reason other than a node along the path was malicious" which are currently considered exactly opposite, but with `23` they aren't anymore.

I gather we treat `23` as payment retryable, right? Since it's not `PERM`, won't it be treated correctly by the existing logic or is there something else I am missing?

https://github.com/rust-bitcoin/rust-lightning/blob/4a374084c4655e0c13d94d799688842278f9a94a/lightning/src/ln/onion_utils.rs#L477

**Comment by TheBlueMatt** [comment]
> Since it's not PERM, won't it be treated correctly by the existing logic or is there something else I am missing?

Yea, the call flow here isn't entirely clear, sorry. So `process_onion_failure`'s third return argument (generally `!is_from_final_node`) ends up being assigned as `payment_retryable` in `ChannelManager::fail_htlc_backwards_internal` which is then assigned as `rejected_by_dest: !payment_retryable` in `Event::PaymentPathFailed`, the docs for which suggest that if you see `rejected_by_dest` the payment is not retryable as the recipient has indicated they do not wish to/are unable to accept the payment. That isn't what we want here, of course - the payment is definitely retryable, and even this path is good, we just timed out. For the current `Event::PaymentPathFailed` API the correct return values, I believe, would be `!payment_failed` and `short_channel_id` of `None` (implying there is no channel to de-prioritize for future pathfinding.

Ultimately the issue here is that the failure modes of an HTLC are not really accurately communicable with a simple boolean in `Event::PaymentPathFailed`. No need to do it here, but in the future we should probably think about moving `rejected_by_dest`, `network_udpate` and `short_channel_id` all into one big `HTLCFailureReason` enum.

**Comment by dunxen** [comment]
Thanks for the context here 🙂
I've got a much clearer picture now.

**Comment by TheBlueMatt** [comment]
Let me know when you want another review pass on this PR!

**Comment by dunxen** [comment]
> Let me know when you want another review pass on this PR!

Thanks! I think it's ready for another pass.

I added an onion route test to try convince myself of something and it passes. So I believe we don't need to do as much refactoring as initially thought right now.

But it seems that any `payment_failed` that is not permanent would be considered retryable currently. Don't think that's what we want (other than for 23) and maybe that's where the problem lies.

Sorry this has taken a little long, have been wrapping up a project at work.

File: lightning/src/ln/msgs.rs [hunk 2]
```diff
@@ -68,6 +68,8 @@ pub enum DecodeError {
         io::ErrorKind),
 	/// The message included zlib-compressed values, which we don't support.
 	UnsupportedCompression,
+	/// The message included a TLV with an unknown type.
+	UnknownTLV(u64),
```
Comment (TheBlueMatt) [inline lightning/src/ln/msgs.rs:72]: IIUC this variant is never created in this PR? Maybe it can be added later if needed when its crated?

File: lightning/src/util/errors.rs [hunk 3]
```diff
@@ -119,6 +119,8 @@ pub(crate) fn get_onion_error_description(error_code: u16) -> (&'static str, &'s
 		_c if _c == 19 => ("The final node indicated the amount in the HTLC does not match the value in the onion", "final_incorrect_htlc_amount"),
 		_c if _c == UPDATE|20 => ("Node indicated the outbound channel has been disabled", "channel_disabled"),
 		_c if _c == 21 => ("Node indicated the CLTV expiry in the HTLC is too far in the future", "expiry_too_far"),
+		_c if _c == PERM|22 => ("Node indicated that the decrypted onion per-hop payload was not understood by it or is incomplete", "invalid_onion_payload"),
```
Comment (TheBlueMatt) [inline lightning/src/util/errors.rs:122]: Should we add a test for 22 as well?

**Comment by TheBlueMatt** [comment]
> But it seems that any payment_failed that is not permanent would be considered retryable currently. Don't think that's what we want (other than for 23) and maybe that's where the problem lies.

From my read its just the PERM && from_last_node that makes it non-retryable, which seems right?

File: lightning/src/ln/msgs.rs (hunk 2 above)
Comment (dunxen) [reply lightning/src/ln/msgs.rs:72]: Yeah, true. I think I left it behind when I added that out of scope stuff. Will remove 🙂

**Comment by dunxen** [comment]
Yeah, that's in the `payment_failed` branch, right? So I'm just thinking if we had a non PERM error that is also one of the `payment_failed` errors, then it would be considered retryable.

**Comment by dunxen** [comment]
Ok wait I convinced myself that 18 and 19 should be retryable since we do attribute blame to a hop, so it is retryable on a different route right?

**Comment by TheBlueMatt** [comment]
> Yeah, that's in the payment_failed branch, right? 

No, the bottom of the `if let Some(error_code_slice) = err_packet.failuremsg.get(0..2) {` branch sets `res`'s last arg to `!(error_code & PERM == PERM && is_from_final_node)`

**Comment by dunxen** [comment]
> > Yeah, that's in the payment_failed branch, right? 
> 
> 
> 
> No, the bottom of the `if let Some(error_code_slice) = err_packet.failuremsg.get(0..2) {` branch sets `res`'s last arg to `!(error_code & PERM == PERM && is_from_final_node)`

Yeah, my brain is on holiday 😂
Thanks!

File: lightning/src/util/errors.rs (hunk 3 above)
Comment (dunxen) [reply lightning/src/util/errors.rs:122]: Seems we already do have this in `onion_route_tests.rs` for intermediate and final node case.

**Review by TheBlueMatt — APPROVED** [review]

**Review by jkczyz — APPROVED** [review]
